src/core/fem/src/discretization/4C_fem_discretization.hpp

```diff
@@ -76,6 +77,217 @@ namespace Core::IO
 
 namespace Core::FE
 {
+
+  class ElementRef;
+  class ConstElementRef;
+
+  /**
+   * Implementation helper for NodeRef and ConstNodeRef.
+   */
+  template <bool is_const>
+  class NodeRefImpl
+  {
+   public:
+    template <typename T>
+    using MaybeConst = std::conditional_t<is_const, const T, T>;
+
+    using ElementRefType = std::conditional_t<is_const, ConstElementRef, ElementRef>;
+
+    //! The discretization type, const or non-const depending on is_const.
+    using DiscretizationType = MaybeConst<Discretization>;
+
+    /**
+     * There is usually no need to call this constructor. You will receive objects of this type
+     * from Discretization methods.
+     */
+    NodeRefImpl(DiscretizationType* dis, int local_id) : discretization_(dis), local_id_(local_id)
+    {
+      FOUR_C_ASSERT(dis, "NodeRefImpl needs a valid discretization");
+    }
+
+    /**
+     * Return the spatial coordinates of the node.
+     */
+    [[nodiscard]] std::span<const double, 3> x() const
+    {
+      const auto& coord = discretization_->nodecolptr_[local_id_]->x();
+      FOUR_C_ASSERT(coord.size() == 3, "NodeRef only supports 3D coordinates");
+      return std::span<const double, 3>(coord.data(), 3);
+    }
+
+    /**
+     * Return the owning MPI rank of the node.
+     */
+    [[nodiscard]] int owner() const { return discretization_->nodecolptr_[local_id_]->owner(); }
+
+    /**
+     * Return the local id of the node in the discretization.
+     */
+    [[nodiscard]] int local_id() const { return local_id_; }
+
+    /**
+     * Return the global id of the node in the discretization.
+     */
+    [[nodiscard]] int global_id() const { return discretization_->nodecolptr_[local_id_]->id(); }
+
+    /**
+     * Return a pointer to user data of type Node. This may be nullptr.
+     */
+    [[nodiscard]] MaybeConst<Nodes::Node>* user_node() const
+    {
+      return discretization_->nodecolptr_[local_id_];
+    }
```

Comment:
Don't you want to have this method non-const if it is returning a non-const pointer?
```suggestion
    [[nodiscard]] MaybeConst<Nodes::Node>* user_node()
    {
      return discretization_->nodecolptr_[local_id_];
    }
    
    [[nodiscard]] const Nodes::Node* user_node() const
    {
      return discretization_->nodecolptr_[local_id_];
    }
```

Reply:
yes, you are right :+1: